Document postponable unattended-upgrades (#271)
canonical/ubuntu-server-documentation: how-to/software/automatic-updates.md

Always keep in mind, however, that specific packages can be blocked from receiving updates. For example, if a particular system runs a critical application that could break if certain libraries on the system are updated, then perhaps an acceptable compromise is to block these library packages from receiving upgrades, instead of disabling the whole feature.

> As a middle-ground solution, `unattended-upgrades` also exposes the ability to postpone impending updates to a later time. Read how to configure this feature in the [Postponable updates](unattended_upgrades_postpone) paragraph.

Reviewer: ```suggestion
As a middle-ground solution, `unattended-upgrades` also exposes the ability to postpone impending updates to a later time. Read how to configure this feature in the {ref}`Postponable updates <unattended_upgrades_postpone>` section.
```

An example of such a Fleet Management software for Ubuntu systems is [Landscape](https://ubuntu.com/landscape).

(unattended_upgrades_postpone)=
## Postponable updates
By default, system updates are applied automatically in the background without any user interaction.
Starting with Ubuntu 25.04, this behaviour can be changed to prompt active users before starting the update, allowing them to postpone the updates to a later time, up to a limited number of days as configured by the system administrator.

> To enable the feature, set the `Unattended-Upgrade::Postpone-For-Days` configuration in `/etc/apt/apt.conf.d/50unattended-upgrades` to a positive number, signifying the number of days that a user of this system is allowed to postpone the automatic updates for. Every time `unattended-upgrades` is run, after determining whether some updates are available to be applied it will prompt the user and shortly wait for an answer. If the user requests to postpone an update, the date and time of this action is recorded so that on future runs `unattended-upgrades` will only prompt the user again if the number of days elapsed from the first postpone request is lower than what is configured in `Unattended-Upgrade::Postpone-For-Days`.

Reviewer: ```suggestion
To enable the feature, set the `Unattended-Upgrade::Postpone-For-Days` configuration in `/etc/apt/apt.conf.d/50unattended-upgrades` to a positive number, signifying the number of days that a user of this system is allowed to postpone the automatic updates for. Every time `unattended-upgrades` is run, after determining whether some updates are available to be applied it will prompt the user and shortly wait for an answer. If the user requests to postpone an update, the date and time of this action is recorded. Then, on future runs, `unattended-upgrades` will only prompt the user again if the number of days since the first postpone request is lower than what is configured in `Unattended-Upgrade::Postpone-For-Days`.
```

The feature is disabled if the number of days is set to 0.

> The `Unattended-Upgrade::Postpone-Wait-Time` configuration allows controlling the amount of time (in seconds) that a user of this system has available to send a postpone request after being prompted. If no postpone request is received within the specified time, then the updates will start being applied as normal.

Reviewer: ```suggestion
The `Unattended-Upgrade::Postpone-Wait-Time` configuration option controls the amount of time (in seconds) that a user has available to send a postpone request after being prompted. If no postpone request is received within the specified time, the updates will start being applied as normal.
```

> The prompting functionality is currently implemented graphically on Ubuntu Desktop by the `update-notifier` program. While the updates are being applied, an icon will also be visible in the system tray area informing the user so that they may know when it is safe to resume critical activities that may be affected by the updates.

Reviewer: ```suggestion
The prompting functionality is implemented graphically on Ubuntu Desktop by the `update-notifier` program. While updates are being applied, an icon is visible in the system tray area informing the user so they know when it is safe to resume critical activities that may be affected by the updates.
```

> On other environments, you may implement your own prompting client by listening for the `AboutToStart` signal on the system bus and send a call to the `Postpone()` method. Read the `/usr/share/dbus-1/interfaces/com.ubuntu.UnattendedUpgrade.Pending.xml` interface specification for more details.

Reviewer: ```suggestion
On other environments, such as Ubuntu Server, you can implement your own prompting client by listening for the `AboutToStart` signal on the system bus and send a call to the `Postpone()` method. Read the `/usr/share/dbus-1/interfaces/com.ubuntu.UnattendedUpgrade.Pending.xml` interface specification for more details.
```

Reviewer: Is that all the documentation that exists about this?

Author: Yes it is.

An example of such a Fleet Management software for Ubuntu systems is [Landscape](https://ubuntu.com/landscape).

(unattended_upgrades_postpone)=
## Postponable updates
By default, system updates are applied automatically in the background without any user interaction.

> Starting with Ubuntu 25.04, this behaviour can be changed to prompt active users before starting the update, allowing them to postpone the updates to a later time, up to a limited number of days as configured by the system administrator.

Reviewer 2: This is rewritten content from the following paragraph. I feel like it rephrases the introduction so we could also just remove it.

> Starting with Ubuntu 25.04, this behaviour can be changed to prompt active users before starting the update, allowing them to postpone the updates to a later time, up to a limited number of days as configured by the system administrator.

Author: > `unattended-upgrades` stops prompting the user for the configured number of days.

That's not what happens, and it is why I was so technical in that other paragraph.

Let's assume that `unattended-upgrades` was configured to run once every day.
And that `Unattended-Upgrade::Postpone-For-Days "2"`.
On the 1st day,  the user is prompted, and chooses to postpone.
On the 2nd day, the user is prompted and chooses to postpone again.
On the 3rd day, the user is not prompted anymore because too many days have passed since the first postpone request, and the upgrade is forcefully started. The counter resets and the user will be allowed to postpone again starting from the 4th day if there are any more updates.

But `unattended-upgrades` can be configured to run with any cadence, so here's another example.
Let's assume that `unattended-upgrades` was configured to run once every 2 days.
And that `Unattended-Upgrade::Postpone-For-Days "3"`.
On the 1st day,  the user is prompted, and chooses to postpone.
On the 3nd day, the user is prompted and chooses to postpone again.
On the 5th day, updates are applied without prompting.

Please help me explain this in writing because you're not the first person who misunderstand the behaviour by my writing :(

> Starting with Ubuntu 25.04, this behaviour can be changed to prompt active users before starting the update, allowing them to postpone the updates to a later time, up to a limited number of days as configured by the system administrator.

Reviewer 2: Ah, I see! I did misunderstand, sorry @aleasto. How about this:

For the configured number of days, the user is prompted to install updates. The user can postpone them. When the configured number of days has passed, `unattended-upgrades` stops prompting the user, and on its next run, it installs the updates automatically.

If `unattended-upgrades` runs daily, it installs the updates on the same day that the postpone period runs out. However, it might also run weekly or at a different frequency, and in that case, it installs the updates with a certain delay after the postpone period.

> Starting with Ubuntu 25.04, this behaviour can be changed to prompt active users before starting the update, allowing them to postpone the updates to a later time, up to a limited number of days as configured by the system administrator.

Reviewer: Here's my suggestion, let me know if this helps?
```
By default, system updates are applied automatically in the background without any user interaction.

Starting with Ubuntu 25.04, a system administrator can allow users to postpone these automatic updates for a limited number of days by setting the `Unattended-Update::Postpone-For-Days` option.

When this option is set, `unattended-upgrade` will run according to the cadence set by the administrator and check for updates. If there are updates available it will notify active users and prompt them to choose if they want to upgrade immediately, or postpone them. For example, if `Unattended-Upgrade::Postpone-For-Days "3"` is set, then the user can postpone upgrades for up to three days. After that, the next time `unattended-updates` runs the user will not be prompted and the upgrades will be applied to the system.
```

> Starting with Ubuntu 25.04, this behaviour can be changed to prompt active users before starting the update, allowing them to postpone the updates to a later time, up to a limited number of days as configured by the system administrator.

Author: Thank you both!

> To enable the feature, set the `Unattended-Upgrade::Postpone-For-Days` configuration in `/etc/apt/apt.conf.d/50unattended-upgrades` to a positive number, signifying the number of days that a user of this system is allowed to postpone the automatic updates for. Every time `unattended-upgrades` is run, after determining whether some updates are available to be applied it will prompt the user and shortly wait for an answer. If the user requests to postpone an update, the date and time of this action is recorded. Then, on future runs, `unattended-upgrades` will only prompt the user again if the number of days since the first postpone request is lower than what is configured in `Unattended-Upgrade::Postpone-For-Days`.

Reviewer 2: ````suggestion
To enable the feature, edit the `/etc/apt/apt.conf.d/50unattended-upgrades` and set the number of days that a user is allowed to postpone the automatic updates for. To postpone for **7** days:

```text
Unattended-Upgrade::Postpone-For-Days "7";
```
````

Reviewer 2: The paragraph was quite difficult to read because it was written from the point of view of the implementation. See the reformulated paragraph above.

Another issue was that the instructions didn't show the configuration syntax, even though it's obvious to experienced APT users. I've added a clear example.

> The feature is disabled if the number of days is set to 0.

Reviewer 2: ```suggestion

To disable the feature, set the number of days to **0**.
```

Reviewer 2: This is just to use the same style for enable and disable.

> The `Unattended-Upgrade::Postpone-Wait-Time` configuration option controls the amount of time (in seconds) that a user has available to send a postpone request after being prompted. If no postpone request is received within the specified time, the updates will start being applied as normal.

Reviewer 2: ```suggestion
### Prompt duration

The `Unattended-Upgrade::Postpone-Wait-Time` configuration option controls the amount of time (in seconds) that a user has available to send a postpone request after being prompted. If no postpone request is received within the specified time, the updates will start being applied as normal.
```

Reviewer 2: In this part, I find subheadings helpful to understand the use cases.

> The system administrator can restrict access to the postpone request by defining Polkit rules for the `com.ubuntu.UnattendedUpgrade.Pending.Postpone` action. By default, access is granted to users of an active session.

Reviewer 2: ```suggestion
### Who can postpone

The system administrator can restrict access to the postpone request by defining Polkit rules for the `com.ubuntu.UnattendedUpgrade.Pending.Postpone` action. By default, access is granted to users of an active session.
```

> 

Reviewer 2: ```suggestion
### Notifications in different environments

```

The `Unattended-Upgrade::Postpone-Wait-Time` configuration option controls the amount of time (in seconds) that a user has available to send a postpone request after being prompted. If no postpone request is received within the specified time, the updates will start being applied as normal.

> The system administrator can restrict access to the postpone request by defining Polkit rules for the `com.ubuntu.UnattendedUpgrade.Pending.Postpone` action. By default, access is granted to users of an active session.

Reviewer: We don't have any documentation in the server docs (currently) about Polkit - do we have anything on the Desktop side that exists that we can point to?

Author: No, but there's https://www.freedesktop.org/software/polkit/docs/latest/polkit.8.html which we could link to?

It has an "AUTHORIZATION RULES" section which is what I mention here.

Author: It's also available as a manpage: https://manpages.ubuntu.com/manpages/plucky/man8/polkit.8.html

Reviewer: Both look like nice references - could you add them both? Then I'll be happy to merge :)

Author: Well it's the same content in both references, except freedesktop.org has images while manpages.ubuntu.com has ASCII art. I don't think referencing both makes sense, then which one would you point to?

Reviewer: Usually the way we do this (we do it quite often) is if there's a web version we provide a hyperlink to the external URL, and if there's a manpage we show the man command.

Reviewer 2: I think that the freedesktop link has more value than the man page because the diagram is more readable and it features two screenshots, too. If we want to add the man page as well, I don't mind, but it only duplicates the information in a less readable format.
